### PR TITLE
Implement job state machine validations

### LINKED ISSUE #12058: [Design/Refactoring] Hide the temporal activity calls behind a state machine implementation
## Tell us about the problem you're trying to solve

It is complicated to properly understand what are the diiferent job's state transition within temporal. Those transition are non explicit in the code and rely on call to the temporal activity. It would be great to abstract that and have a workflow totally un-correlated to temporal. 

Proposal:
One interface and its implementation that will do a job state transition, from an original state and some internal state (Did the job ran already, what is the status of the sync). For example, if we call this method from a with an orginal state of RUNNING and with a sync being FAILED, it will make the transition of the state from RUNNING to FAILED. This need to be abstracted in order to be able to call this interface without knowing that temporal is running it behind the scene. It should also know how to run the sync.

### PR DESCRIPTION
**Pull request by lmossman**
## What
Resolves https://github.com/airbytehq/airbyte/issues/12058

After discussions in backlog grooming on the above ticket, we had concluded that the job status transitions were encapsulated fairly well in the `JobCreationAndStatusUpdateActivityImpl` class, but we still weren't codifying the job state machine in our docs anywhere, so we wanted to add that validation logic.

This PR accomplishes that goal by adding a simple HashMap representation of the allowed job status transitions, and enforcing this in the DefaultJobPersistence class when performing job status transitions.

## How
One callout is that the job state machine shown in our [current documentation](https://docs.airbyte.com/understanding-airbyte/jobs#job-state-machine) is not totally accurate in terms of the status transitions that actually happen today in the new temporal scheduler workflow. The main changes are:
1. Any non-terminal job status can be transitioned to `FAILED`, as this is what happens when we restart a temporal workflow from a clean state.
2. When an attempt fails and the parent job is transitioned to `INCOMPLETE`, the old scheduler had to first reset the job to `PENDING` before creating a new attempt causing it to be set to `RUNNING`. In the new temporal scheduler, jobs are set to `INCOMPLETE` when an attempt fails, and then the temporal workflow loops back around and creates a new attempt, causing the job to move from `INCOMPLETE` to `RUNNING`.

Since point 1 was a change we intentionally made, and point 2 didn't feel worth changing our temporal workflow logic to match the old state machine, I instead opted to just update our state machine diagram to reflect the new state machine that is in effect today/

## Recommended reading order
1. `JobStatus.java`
2. `Job.java`
3. `DefaultJobPersistence.java`
4. `job-state-machine.png` - this is the updated job state machine diagram

## 🚨 User Impact 🚨
There should be no user impact, since this is just codifying the state transitions that we are performing today.

### CODE REVIEW

**Review by jdpgrailsdev — APPROVED** [review]
:shipit:

**Review by cgardens — APPROVED** [review]
nice!

File: airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java [hunk 1]
```diff
@@ -19,4 +20,12 @@ public enum JobStatus {
   public static final Set<JobStatus> TERMINAL_STATUSES = Sets.newHashSet(FAILED, SUCCEEDED, CANCELLED);
   public static final Set<JobStatus> NON_TERMINAL_STATUSES = Sets.difference(Set.of(values()), TERMINAL_STATUSES);
 
+  public static final Map<JobStatus, Set<JobStatus>> VALID_STATUS_CHANGES = Map.of(
+      PENDING, Set.of(RUNNING, FAILED, CANCELLED),
+      RUNNING, Set.of(INCOMPLETE, SUCCEEDED, FAILED, CANCELLED),
+      INCOMPLETE, Set.of(PENDING, RUNNING, FAILED, CANCELLED),
+      SUCCEEDED, Set.of(),
+      FAILED, Set.of(SUCCEEDED),
```
Comment (benmoriceau) [inline airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java:28]: Can we go from failed/cancelled to succeeded? I though there were terminal. I saw the graph and I don't think it should happen

File: airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java [hunk 2]
```diff
@@ -19,4 +20,12 @@ public enum JobStatus {
   public static final Set<JobStatus> TERMINAL_STATUSES = Sets.newHashSet(FAILED, SUCCEEDED, CANCELLED);
   public static final Set<JobStatus> NON_TERMINAL_STATUSES = Sets.difference(Set.of(values()), TERMINAL_STATUSES);
 
+  public static final Map<JobStatus, Set<JobStatus>> VALID_STATUS_CHANGES = Map.of(
```
Comment (benmoriceau) [inline airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java:23]: This is great!

**Comment by benmoriceau** [comment]
I have added @davinchia to the review because this change comes from a discussion we had with him.

File: airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java (hunk 1 above)
Comment (davinchia) [reply airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java:28]: same thought. If `cancelled`, `failed` and `succeeded` are terminal, I'd think that set of valid transitions should be null.

**Review by davinchia — COMMENTED** [review]
Definitely an improvement! One questions from me on the transitions before approving.

File: airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java (hunk 1 above)
Comment (lmossman) [reply airbyte-scheduler/models/src/main/java/io/airbyte/scheduler/models/JobStatus.java:28]: Took another look at this, and I agree with both of you. These were status transitions leftover from the old scheduler days, and it doesn't look like they are able to occur now in the temporal scheduler. I will remove these

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/activities/JobCreationAndStatusUpdateActivityTest.java [hunk 3]
```diff
@@ -290,9 +290,9 @@ public void setAttemptFailureWrapException() throws IOException {
     public void setJobCancelled() throws IOException {
       jobCreationAndStatusUpdateActivity.jobCancelled(new JobCancelledInput(JOB_ID, ATTEMPT_ID, failureSummary));
 
-      Mockito.verify(mJobPersistence).cancelJob(JOB_ID);
       Mockito.verify(mJobPersistence).failAttempt(JOB_ID, ATTEMPT_ID);
       Mockito.verify(mJobPersistence).writeAttemptFailureSummary(JOB_ID, ATTEMPT_ID, failureSummary);
+      Mockito.verify(mJobPersistence).cancelJob(JOB_ID);
```
Comment (benmoriceau) [inline airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/activities/JobCreationAndStatusUpdateActivityTest.java:295]: Nit, is it testing the order or should we add is explicitely using InOrder: https://stackoverflow.com/questions/21901368/mockito-verify-order-sequence-of-method-calls

Comment (lmossman) [reply airbyte-workers/src/test/java/io/airbyte/workers/temporal/scheduling/activities/JobCreationAndStatusUpdateActivityTest.java]: Wasn't aware of that, thanks for the suggestion! I added that to this test

**Pull request merged**